hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/NodeEndpoint.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.ozone.recon.api;
+
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.protocol.proto.HddsProtos.NodeState;
+import org.apache.hadoop.hdds.scm.container.placement.metrics.SCMNodeStat;
+import org.apache.hadoop.hdds.scm.node.states.NodeNotFoundException;
+import org.apache.hadoop.hdds.scm.pipeline.PipelineID;
+import org.apache.hadoop.hdds.scm.server.OzoneStorageContainerManager;
+import org.apache.hadoop.ozone.recon.api.types.DatanodeMetadata;
+import org.apache.hadoop.ozone.recon.api.types.DatanodeStorageReport;
+import org.apache.hadoop.ozone.recon.api.types.DatanodesResponse;
+import org.apache.hadoop.ozone.recon.scm.ReconNodeManager;
+
+import javax.inject.Inject;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Endpoint to fetch details about datanodes.
+ */
+@Path("/datanodes")
+@Produces(MediaType.APPLICATION_JSON)
+public class NodeEndpoint {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(NodeEndpoint.class);
+
+  private ReconNodeManager nodeManager;
+
+  @Inject
+  NodeEndpoint(OzoneStorageContainerManager reconSCM) {
+    this.nodeManager =
+        (ReconNodeManager) reconSCM.getScmNodeManager();
```

Comment:
Is there any reason to inject `OzoneStorageContainerManager` instead of `ReconNodeManager` (just curious).

Reply:
@elek Thanks for reviewing this. We cannot inject 'ReconNodeManager' since it is not part of the injected classes in the Controller Module. If we bring in the node manager through injection, we will have to move all the other sub modules within Recon SCM also into injection. Currently, they are instantiated very similar to the way in actual SCM, staying true to the "Recon as Passive SCM" story.

Reply:
Thanks the answer. I had similar suspicion. IMHO even without dependency injection we can do better separation between creating and wiring the objects on the SCM side. For example we can do the wiring on the SCM side in a separated utility and in this case you can do better wiring on recon.

Or even we can introduce dependency injection on the SCM side as well.

(Not in this patch. Just thinking loudly ;-))